### DISCUSSION TITLE
Warning: ST7796 and ILI9488 TFT with touch controller

### DISCUSSION THREAD

**Discussion by Bodmer** [Bodmer/TFT_eSPI#898] [Alerts]
### ST7796 and ILI9486 displays
On some ST7796 and ILI9486 boards the SPI output ( SDO/MISO) pin does not go tristate in certain circumstances because a diode is fitted (see below) in the chip select line. This is a problem if other SPI devices occupy the bus (e.g. if an SD card or touch controller is used) for two reasons:
1. The SDO/MISO pin may not go tristate when the TFT chip select is high (it depends on the direction of parasitic input currents).
2. Even if the TFT SDO/MISO pin is disconnected, bus transactions to other devices such as the touch controller may still corrupt the displayed image. (because the rise time of the chip select is so slow).

The correct solution when other devices will occupy the SPI bus is to short/remove the diode, see next comment.

### ILI9488 displays
The ILI9488 SPI output ( SDO/MISO) does not go tristate and this cannot be circumvented, so do NOT connect the display interface SDO/MISO pin to the processor if other SPI devices are on the same SPI bus. You can then connect other SPI devices to the processor MISO pin such as the touch controllers DO (Data Out) pin.

**Comment by Bodmer** [comment]
Some ST7796 (and ILI9486) displays have a diode inline with the TFT chip select. This is a board design error as there is nothing the other side of the diode to pull the TFT chip select high. If there are other devices on the SPI bus, such as a touch controller then this diode is likely to cause problems.

Diode location:
![Inked20201208_113227_LI](https://user-images.githubusercontent.com/15824805/103423976-c319d500-4ba1-11eb-90d2-c3993ef91053.jpg)

Unfortunately the solution requires soldering skills, remove the diode and replace with a wire link  or add a shorting link over the diode:

![TFT_CS link](https://user-images.githubusercontent.com/15824805/103424030-07a57080-4ba2-11eb-8b9c-af7de838e93e.jpg)

**Reply by FIyingcar** [reply]
非常感谢

**Comment by Bodmer** [comment]
[Here is the schematic](http://www.lcdwiki.com/res/MSP4021/4.0inch_SPI_Schematic.pdf) for those rogue boards with a diode fitted:
![tft_diode](https://user-images.githubusercontent.com/15824805/103424364-9e266180-4ba3-11eb-952e-75415e5648be.png)

**Comment by alesimattia** [comment]
Thank you very much, man!

**Comment by mstrens** [comment]
Thank you very much for the library.
I am facing issues with this TFT too.
I am blocked because I have the same TFT as on the photo. I have to connect it to a pcb that use the same SPI for the display and for the touchscreen. I am using a ESP32 but I am not sure it is meaningful.
When I get the display working, the touchscreen does not work.
I already replace the diode by a wire but it did not solved the issue.
I wrote special sketch (without using your librairy) to test what happens .
I noticed that the issue is related to the TFT_RST pin.
When the TFT_RST pin is defined as output and set to HIGH, the XPT2046 returns only a value equal to zero when I read it. Otherwise, it replies with values that seems OK (they change when I touch the screen).
I have no solution for this issue.
I checked with an ohmmeter for any unexpected connection between the TFT reset pin and other tft pins but I did not find one.
 
Note: I have a friend that has the same TFT. He did not remove the diode (was not aware off this) and his TFT works just fine using exactly the same firmware and the same board.

Very strange.
I already looked for more than one day and do not understand what happens.
I will see if I can cut the tft MISO wire on the TFT but I would like to know if you have another idea.

**Reply by eagl1** [reply]
I think you should solder a 4.7k resistor between the RST and the VCC. And connect the RST to the controller, as the library manages the timing to pull the RST pin to LOW state.

**Comment by mstrens** [comment]
I tried to remove the pin MISO from the TFT connector and now I get both the TFT and the touch screen working.
I let this message here because it can be useful for other users.

**Comment by Bodmer** [comment]
Glad to hear it is working for you.  To answer some of the points raised:
1. When the TFT_RST is high the TFT chip will be active and could drive the MISO line. When low the TFT chip is being reset and lines go tr-state (high impedance).
2. Removing the diode and fitting a link should work. When TFT_CS is low the MISO line will be driven by the display chip, when high it should not be, so make sure it is high when the touch controller is being selected (TOUCH_CS driven low).
3. If your friend has the diode fitted, then he may be one of the lucky ones where the TFT_CS on the other side of the diode (display chip side) floats high when the TFT_CS line is high. The diode prevents TFT_CS driving the dispay side signal high, but it can naturally drift high. When the line drifts high this can take an appreciable time (miliseconds) so writing to the display then immediately reading the touch controller may give problems (bad touch readings and corrupted pixels on display).

The TFT MISO signal track is visible on the screen side of the display PCB so it could be cut there, or the pin unsoldered and pulled out,

**Reply by Aleks-Ale** [reply]
> The TFT MISO signal track is visible on the screen side of the display PCB so it coule be cut there ot the pin unsoldered and pulled out

Is there any suggestion where there may be such a short circuit of the MISO to the ground? 
The visible part of the signal track is too short to solve the problem. A short circuit is also possible in the display panel.

**Reply by Bodmer** [reply]
Cutting the track has no benefit but was suggested by @mstrens to solve a problem. I just pointed out where this track can be found.

The correct solution is to removes the diode and/or add a shorting link as detailed in the pictures. This then makes the display behave "normally", so when TFT_CS is high the MISO line tri-states.

**Reply by Aleks-Ale** [reply]
> The correct solution is to removes the diode and/or add a shorting link as detailed in the pictures.

I did it yesterday. Unfortunately, this did not solve the problem. If the display MISO is connected, other SPI devices do not work. For example, VS1053. If the display MISO is not connected, the display does not show the correct picture. That's why I'm wondering where a short circuit is possible to make a new connection to the pin.

**Reply by Aleks-Ale** [reply]
> Cutting the track has no benefit but was suggested by @mstrens to solve a problem. I just pointed out where this track can be found.

You're both right. For example, interlayer short circuits in boards are repaired like this.

**Reply by Bodmer** [reply]
There is no short circuit. To be clear the diode can prevent the display controller chip select being driven high, so then the display is enabled.

It TFT_CS is driven high then the display SDO output should go tri-state. Other users have found this to be the case but I have just received an ST7796 touch screen display so will try it.

You only need to connect the display SDO (MISO) line if you wish to read from the display. If it is connected to the processor and the display is read then pixel data will be received correctly, so there is NO short circuit on the track.

**Reply by mecatronicaindustrialmaker** [reply]
Esou passando pelo mesmo problema e não estou conseguindo

**Comment by mstrens** [comment]
Thanks for your quick answer.
It is clear.
Still I am not sure why removing the diode did not fix the issue for me but it could be that, in my own skectch to read the touchscreen, I had the pin TFT_RST forced to HIGH but I was not forcing a level High on TFT_CS (It was probably floating).

**Comment by mstrens** [comment]
I do not understand why the display should not show correct picture when the TFT MISO pin is not connected.
It is an output signal from display controller so why should it have an impact if the MCU does not have to read the display RAM.
I presume (but did not checked) that the library reads the TFT RAM only for advanced functions.

**Reply by Bodmer** [reply]
The problem should only occur if the TFT chip slect line cannot be pulled high, which can be the case when the diode is fitted.

If the TFT chip select stays low, then writes to the display work normally, however if it stays low while writes to other SPI devices occur on the MOSI line (for example writes to the XPT2046 touch controller occur to tell it to respond with touch data) then that data destined to the other device is liklely to appear on the display as bad pixels, these will be in the last window area accessed on the TFT.

**Comment by mstrens** [comment]
I would like to develop a board based on ESP32 and a 4" ST7796 TFT with touch screen.
It would be to difficult to ask the users to remove the diode on CS line.
In order to use the display without removing the diode, I imagine it should be possible to use one ESP32 SPI (e.g. HSPI) to manage the TFT (ST7796) and another SPI (e.g. VSPI) to manage the touchscreen.
I presume that this possibility has not been foreseen in your library.
Do you think it should be difficult to add it?

**Reply by Bodmer** [reply]
The library asumes that the touch controller is on the same SPI pins, so if you wish to use the other port you will need to use an independant library. There are a few such as https://github.com/achillhasler/TFT_eTouch, or https://github.com/PaulStoffregen/XPT2046_Touchscreen.

**Reply by baronengineer** [reply]
Can this library be used with a resistive ILI9488 TFT?

**Comment by bruno13660** [comment]
hello,
I should  want use a TFT 4" display with ILI7796S and use SD card  on the same SPI connection . 
This post is old ( March 2020 )...; Must I short always the diode on Cs .?
Does Some body try to put a pull resistor befor the diode ? 
Thank you for the help ..;
bruno ( I am french )

**Comment by Bodmer** [comment]
You could  try a pull up on the Anode (display side) of the diode. This would require modification to the display PCB.

**Comment by vick3394** [comment]
Mr Bodmer.
Thank you for your patience.
![unnamed](https://user-images.githubusercontent.com/72776429/188040843-a855ad97-d213-4fb7-86b5-c61909929eff.jpg)

I have a 4 inch TFT LCD running on driver ILI9488. I use a ESP32
Attached is the wiring on my PCB. It looks different from the above. 
I do not wish to use the Touch screen. I just need my SD card to work to run a slide show using the Xtronical program.
Do I still have to remove the diode?
I am able to run the example on the Arduino IDE TFT_meters. But when I run the Xtronical program, It cannot read the JPEG files in my SD card. Do I have to use a separate SDcard reader?

Thank you.

**Reply by tof80520** [reply]
Bonjour à Tous,
je ne parviens pas à utiliser la bibliothèque avec un esp32-3248S035C, cela vient-il de ce problème?
Merci à vous.

**Reply by Bodmer** [reply]
The problem is with a particular display design. It is unlikely this problem has been duplicated in that board.

**Comment by Rob58329** [comment]
I have “4.0 TFT SPI 480x320 v1.1 (SP7796)" which has a different D1 location, BUT it looks like the schematic is still the same (ie. D1 is still in the CS line), and it needs to be shorted out for multiple devices on the SPI bus to work properly.  The photo below shows my v1.1 board with a wire/solder-blob shorting out D1.


![v1 1_D1_shorted](https://user-images.githubusercontent.com/30350495/205449084-23e96734-ac23-42cc-a927-c7a2a862ef59.jpg)

[The only reason for D1 that I can think of would be if the matrix_CS (via the 48pin FPC) needs different voltage levels to 3.3V.  The matrix_CS must also have an internal pullup (not shown on schematic), otherwise my simultaneously connected SPI-SD card would not half-work with D1 in.  But this internal pullup must be really weak as removing it significantly improves the matrix_CS-high(disable)speed.  It’s interesting to note that the new board revision v1.1 seems to be the same schematic as the original v1.0, so D1 does not seem to be included in error??  Perhaps 2 parallel Diodes connected in opposite directions would be a better alternative than shorting-out, but harder to modify to this.]

**Reply by Bodmer** [reply]
The CS line accepts the same input voltages as the MISO and SCK pins. There are ST7796 boards that do not have this diode in the circuit. The designer probably had a reason to fit the diode, maybe for a particular product with a higher voltage CS control signal? 

Data sheet here has the electrical spec.:
https://www.crystalfontz.com/controllers/Sitronix/ST7796S/

These boards are fitted with a 3.3V regulator (typically labelled U1) sp VCC can be a 5V supply providing J1 is not shorted to bypass the regulator. The ST7796 accepts 3.3V logic levels.

**Reply by Rod581958** [reply]
I have the same TFT SPI display the 480x320 V1.1 with the screen all white or blinking white. This on a Arduino Mega. Did you find the shorting D1 out worked at all and were you getting any correct display from the card.

**Reply by alseenwalnoot** [reply]
> I have the same TFT SPI display the 480x320 V1.1 with the screen all white or blinking white. This on a Arduino Mega. Did you find the shorting D1 out worked at all and were you getting any correct display from the card.
> 
> 

What touch screen do you use? You should try without touch first if you have connected it first. I recommend searching on YouTube for  your exact setup, and you shouldnt short D1. Try connecting to the main spi pins that arduino you should connect to those first and then define them in your setup.

**Reply by Rod581958** [reply]
I have tried 4.0inch_SPI_Module_ST7796S_MSP4020&MSP4021_V1.0, Adafruit_ILI9341, Adafruit_TouchScreen, ILI9341_t3n, ILI9486_SPI-1.0.5, MCUFRIEND_kbv just toname a few. Mine is the exact same one as in tnhe photo. This one of the pinouts for the display.
The connections:
TFT // Mega
VCC // 5V
GND // GND
CS // A5
Reset // A4
DC/CD // A3
SDI(MOSI) // 51
SCK // 52
LED // A0
SDO (MISO) // 50

**Comment by RafaelxVS** [comment]
Hi,
Please, I need some advice.
I'm using the exact same display model (ST7796 V1.0) and an Arduino Mega2560.
I unsoldered D1 and jumped the contacts like the instructions above.
I can't do the things work.
The connections:
TFT  //  Mega
VCC // 5V
GND // GND
CS   // A5
Reset // A4
DC/CD  // A3
SDI(MOSI) // 51
SCK   //  52
LED  //  A0
SDO (MISO) // 50

Attached my User_Setup and User_Select.
[TFT_eSPI.zip](https://github.com/Bodmer/TFT_eSPI/files/10548543/TFT_eSPI.zip)

Please, can anyone help with this issue?

Thanks!
Regars.

**Reply by tvixen** [reply]
Just so you're clear, the Mega2560 is 5V logic and your display is 3V logic. So you can burn your tft display very easily.
You should use level shifters like TXS0108E.  And please disconnect MISO, as written in the tread.

**Comment by baronengineer** [comment]
I have a 3.5-inch resistive ILI9488 TFT. When I input the code with the FT6236 library, the touchscreen does not function. I would appreciate your assistance.

**Comment by Cal6227** [comment]
I'm having problems with the 4 inch resistive TFT V1.1 (driver ST7796) working with a SD card on ESP32.  The D1 jumper is implemented.

SD connections (VSPI): 
#define SD_CS 5
#define SCLK 18  
#define MISO 19  
#define MOSI 23  

TFT connections (HSPI):
#define TFT_MISO 12   
#define TFT_MOSI 13
#define TFT_SCLK 14
#define TFT_CS   15  
#define TFT_DC    2  
//#define TFT_RST   4  // Reset pin (could connect to RST pin)
#define TFT_RST  -1  // Set TFT_RST to -1 if display RESET is connected to 3.3V
#define TOUCH_CS 33    

I'm using SD.h for the SD card library.  

During Setup():
1.  TFT touch works if: SD.begin(SD_CS) is started  before  tft.init()
 SD card can be accessed before TFT is initialized, but not after TFT is initialized.

2.  TFT touch does not work if:  SD.begin(SD_CS) is started  after  tft.init()
In this case SD card can be accessed.

Help in resolving this issue is greatly appreciated.

**Reply by tvixen** [reply]
Please **read** the issue template, under the .github folder. And raise issues only according to the provided examples, as nobody understand half of what you are posting.

To minimise effort to resolve issues the following should be provided as a minimum:
1. A description of the problem and the conditions that cause it to occur
2. IDE (e.g. Arduino or PlatformIO)
3. TFT_eSPI library version (try the latest, the problem may have been resolved!) from the Manage Libraries... menu
4. Board package version (e.g. 2.0.3) available from the Boards Manager... menu
5. Procesor, e.g RP2040, ESP32 S3 etc
6. TFT driver (e.g. ILI9341), a link to the vendors product web page is useful too.
7. Interface type (SPI or parallel)

Plus further information as appropriate to the problem:
1. TFT to processor connections used
2. A zip file containing your setup file (just drag and drop in message window - do not paste in long files!)
3. A zip file  containing a simple and complete example sketch that demonstrates the problem but needs no special hardware sensors or libraries.
4. Screen shot pictures showing the problem  (just drag and drop in message window)

And please make your own discussion thread, instead of stealing other people's threads.

**Comment by jainprince1995** [comment]
My ILI9488 display circuit is little bit different as shown below when I compare it with [LCD-Wiki](http://www.lcdwiki.com/3.5inch_SPI_Module_ILI9488_SKU:MSP3520). 

![IMG_3277](https://github.com/Bodmer/TFT_eSPI/assets/49722562/5a1ccbbe-e50d-41ec-9950-2867352bd599)


To make it running with Arduino Uno, I follow the instruction from [here](http://www.lcdwiki.com/Run_Arduino_Demo_in_spi_model) and
to follow the short circuit method, I couldn't see **_J1_** component on display circuit. Instead of **_J1_** I short **_R0_** to make it running with Uno 5 volt output.

Below it the simple program which I run but I couldn't see anything on display.

```
#define LED   A0    
#define CS    A5        
#define RS    A3       
#define RESET A4
#define MOSI  11
#define SCK   13

void Lcd_Writ_Bus(unsigned char d)
{
   uint8_t val = 0x80;
    while(val)
    {
      if(d&val)
      {
        digitalWrite(MOSI,HIGH);
      }
      else
      {
        digitalWrite(MOSI,LOW);
      }
      digitalWrite(SCK,LOW);
      digitalWrite(SCK,HIGH);
      val >>= 1;
    }
}

void Lcd_Write_Com(unsigned char VH)  
{   
  *(portOutputRegister(digitalPinToPort(RS))) &=  ~digitalPinToBitMask(RS);//LCD_RS=0;
  Lcd_Writ_Bus(VH);
}

void Lcd_Write_Data(unsigned char VH)
{
  *(portOutputRegister(digitalPinToPort(RS)))|=  digitalPinToBitMask(RS);//LCD_RS=1;
  Lcd_Writ_Bus(VH);
}

void Lcd_Write_Com_Data(unsigned char com,unsigned char dat)
{
  Lcd_Write_Com(com);
  Lcd_Write_Data(dat);
}

void Address_set(unsigned int x1,unsigned int y1,unsigned int x2,unsigned int y2)
{
        Lcd_Write_Com(0x2a);
	Lcd_Write_Data(x1>>8);
	Lcd_Write_Data(x1);
	Lcd_Write_Data(x2>>8);
	Lcd_Write_Data(x2);
        Lcd_Write_Com(0x2b);
	Lcd_Write_Data(y1>>8);
	Lcd_Write_Data(y1);
	Lcd_Write_Data(y2>>8);
	Lcd_Write_Data(y2);
	Lcd_Write_Com(0x2c); 							 
}

void Lcd_Init(void)
{
  digitalWrite(RESET,HIGH);
  delay(5); 
  digitalWrite(RESET,LOW);
  delay(15);
  digitalWrite(RESET,HIGH);
  delay(15);

  digitalWrite(CS,LOW);  //CS

    Lcd_Write_Com(0xF7);  
    Lcd_Write_Data(0xA9); 
    Lcd_Write_Data(0x51); 
    Lcd_Write_Data(0x2C); 
    Lcd_Write_Data(0x82);  

    Lcd_Write_Com(0xC0);  
    Lcd_Write_Data(0x11); 
    Lcd_Write_Data(0x09); 

    Lcd_Write_Com(0xC1);  
    Lcd_Write_Data(0x41); 

    Lcd_Write_Com(0xC5);  
    Lcd_Write_Data(0x00); 
    Lcd_Write_Data(0x0A); 
    Lcd_Write_Data(0x80);
 
    Lcd_Write_Com(0xB1);  
    Lcd_Write_Data(0xB0); 
    Lcd_Write_Data(0x11); 

    Lcd_Write_Com(0xB4);  
    Lcd_Write_Data(0x02); 
  
    Lcd_Write_Com(0xB6);    
    Lcd_Write_Data(0x02);
    Lcd_Write_Data(0x22);  
 
    Lcd_Write_Com(0xB7);    
    Lcd_Write_Data(0xC6);  

    Lcd_Write_Com(0xBE);    
    Lcd_Write_Data(0x00);   
    Lcd_Write_Data(0x04); 
 
    Lcd_Write_Com(0xE9);    
    Lcd_Write_Data(0x00);   
 
    Lcd_Write_Com(0x36);  
    Lcd_Write_Data(0x08);   

    Lcd_Write_Com(0x3A);    
    Lcd_Write_Data(0x66); 

    Lcd_Write_Com(0xE0);    
    Lcd_Write_Data(0x00);  
    Lcd_Write_Data(0x07); 
    Lcd_Write_Data(0x10); 
    Lcd_Write_Data(0x09); 
    Lcd_Write_Data(0x17); 
    Lcd_Write_Data(0x0B); 
    Lcd_Write_Data(0x41); 
    Lcd_Write_Data(0x89); 
    Lcd_Write_Data(0x4B); 
    Lcd_Write_Data(0x0A); 
    Lcd_Write_Data(0x0C); 
    Lcd_Write_Data(0x0E); 
    Lcd_Write_Data(0x18); 
    Lcd_Write_Data(0x1B); 
    Lcd_Write_Data(0x0F); 

    Lcd_Write_Com(0xE1);    
    Lcd_Write_Data(0x00);  
    Lcd_Write_Data(0x17); 
    Lcd_Write_Data(0x1A); 
    Lcd_Write_Data(0x04); 
    Lcd_Write_Data(0x0E); 
    Lcd_Write_Data(0x06); 
    Lcd_Write_Data(0x2F); 
    Lcd_Write_Data(0x45); 
    Lcd_Write_Data(0x43); 
    Lcd_Write_Data(0x02); 
    Lcd_Write_Data(0x0A); 
    Lcd_Write_Data(0x09); 
    Lcd_Write_Data(0x32); 
    Lcd_Write_Data(0x36); 
    Lcd_Write_Data(0x0F); 

    Lcd_Write_Com(0x11);    //Exit Sleep 
    delay(120); 			
    Lcd_Write_Com(0x29);    //Display on 

    digitalWrite(CS,HIGH);
}

void H_line(unsigned int x, unsigned int y, unsigned int l, unsigned int c)                   
{	
  unsigned int i,j;
  digitalWrite(CS,LOW);
  Lcd_Write_Com(0x02c); //write_memory_start
  //digitalWrite(RS,HIGH);
  l=l+x;
  Address_set(x,y,l,y);
  j=l*2;
  for(i=1;i<=j;i++)
  {
      Lcd_Write_Data((c>>8)&0xF8);
      Lcd_Write_Data((c>>3)&0xFC);
      Lcd_Write_Data(c<<3);
  }
  digitalWrite(CS,HIGH);   
}

void V_line(unsigned int x, unsigned int y, unsigned int l, unsigned int c)                   
{	
  unsigned int i,j;
  digitalWrite(CS,LOW);
  Lcd_Write_Com(0x02c); //write_memory_start
  //digitalWrite(RS,HIGH);
  l=l+y;
  Address_set(x,y,x,l);
  j=l*2;
  for(i=1;i<=j;i++)
  { 
      Lcd_Write_Data((c>>8)&0xF8);
      Lcd_Write_Data((c>>3)&0xFC);
      Lcd_Write_Data(c<<3);
  }
  digitalWrite(CS,HIGH);   
}

void Rect(unsigned int x,unsigned int y,unsigned int w,unsigned int h,unsigned int c)
{
  H_line(x  , y  , w, c);
  H_line(x  , y+h, w, c);
  V_line(x  , y  , h, c);
  V_line(x+w, y  , h, c);
}

void Rectf(unsigned int x,unsigned int y,unsigned int w,unsigned int h,unsigned int c)
{
  unsigned int i;
  for(i=0;i<h;i++)
  {
    H_line(x  , y  , w, c);
    H_line(x  , y+i, w, c);
  }
}

int RGB(int r,int g,int b)
{
  return r << 16 | g << 8 | b;
}

void LCD_Clear(unsigned int j)                   
{	
  unsigned int i,m;
  digitalWrite(CS,LOW);
  Address_set(0,0,320,480);
  for(i=0;i<320;i++)
    for(m=0;m<480;m++)
    {
      Lcd_Write_Data((j>>8)&0xF8);
      Lcd_Write_Data((j>>3)&0xFC);
      Lcd_Write_Data(j<<3);
    }
  digitalWrite(CS,HIGH);   
}

void setup()
{
  pinMode(A0,OUTPUT);
  pinMode(A3,OUTPUT);
  pinMode(A4,OUTPUT);
  pinMode(A5,OUTPUT);
  pinMode(11,OUTPUT);
  pinMode(13,OUTPUT);

  digitalWrite(A0, HIGH);
  digitalWrite(A3, HIGH);
  digitalWrite(A4, HIGH);
  digitalWrite(A5, HIGH);
  digitalWrite(11, HIGH);
  digitalWrite(13, HIGH);
  
  Lcd_Init();
  
}

void loop()
{  
   LCD_Clear(0xf800);
   LCD_Clear(0x07E0);
   LCD_Clear(0x001F);
   LCD_Clear(0x0); 
  for(int i=0;i<500;i++)
  {
    Rect(random(300),random(300),random(300),random(300),random(65535)); // rectangle at x, y, with, hight, color
  }
  
//  LCD_Clear(0xf800);
}
```

**Comment by onlystudyy** [comment]
hello, I have two espi ili9341 tfts, but they have different problems, one tft can display an image but the touch screen functions by itself without being touched, the second tft can't even display any image, it only shows dim light, is there a problem with pcb board, please anyone who can help
![17211759193936584047523945864065](https://github.com/user-attachments/assets/aa3640c4-026c-4eb6-a7b1-3131e0480125)
![1721175990868752118625844744133](https://github.com/user-attachments/assets/48cca7ce-0b2d-43e1-a398-3a8a50c358b8)

**Comment by alseenwalnoot** [comment]
Hello, this thread is quite old but i still have a question, i have the ILI9488, which works normally, with no touch tho. What differs from the rest is that: i can use the display without the MISO-SDO connection, i can have that connection and the TCS pin and the display still works (no touch), the problem is just that, no matter how much i try i wont get touch to work, i get to the calibration screen, and then when i try to touch it nothing happens

What i've tried: finding the diode (didnt have one), connecting the TDO pin to MISO (where the old SDO would go), finding another library (doesnt exist). I really really need touch so please is there a solution to get it working
